src/main/scala/gitbucket/core/controller/RepositoryViewerController.scala

```diff
@@ -1110,26 +1135,88 @@ trait RepositoryViewerControllerBase extends ControllerBase {
+                val mode = treeWalk.getFileMode.getBits
+                val entry: ArchiveEntry = entryCreator(entryPath, size, mode)
+                JGitUtil.openFile(git, repository, revCommit.getTree, treeWalk.getPathString) { in =>
+                  archive.putArchiveEntry(entry)
+                  IOUtils.copy(in, archive)
+                  archive.closeArchiveEntry()
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+    val tarRe = """\.tar\.(gz|bz2|xz)$""".r
 
-    using(Git.open(getRepositoryDir(repository.owner, repository.name))) { git =>
-      val oid = git.getRepository.resolve(revision)
-      val revCommit = JGitUtil.getRevCommitFromId(git, oid)
-      val sha1 = oid.getName()
-      val repositorySuffix = (if (sha1.startsWith(revision)) sha1 else revision).replace('/', '-')
-      val filename = repository.name + "-" + repositorySuffix + suffix
-
-      contentType = "application/octet-stream"
-      response.setHeader("Content-Disposition", s"attachment; filename=${filename}")
-      response.setBufferSize(1024 * 1024);
-
-      git.archive
-        .setFormat(suffix.tail)
-        .setPrefix(repository.name + "-" + repositorySuffix + "/")
-        .setTree(revCommit)
-        .setOutputStream(response.getOutputStream)
-        .call()
+    filename match {
+      case name if name.endsWith(".zip") =>
+        response.setHeader("Content-Disposition", s"attachment; filename=${filename}")
+        contentType = "application/octet-stream"
+        response.setBufferSize(1024 * 1024);
+        using(new ZipArchiveOutputStream(response.getOutputStream)) { zip =>
+          archive(".zip", zip) { (path, size, mode) =>
+            val entry = new ZipArchiveEntry(path)
+            entry.setSize(size)
+            entry.setUnixMode(mode)
+            entry
+          }
+        }
+      case tarRe(compressor) =>
+        response.setHeader("Content-Disposition", s"attachment; filename=${filename}")
+        contentType = "application/octet-stream"
+        response.setBufferSize(1024 * 1024)
+        using(compressor match {
+          case "gz"  => new GzipCompressorOutputStream(response.getOutputStream)
+          case "bz2" => new BZip2CompressorOutputStream(response.getOutputStream)
+          case "xz"  => new XZCompressorOutputStream(response.getOutputStream)
+        }) { compressorOutputStream =>
+          using(new TarArchiveOutputStream(compressorOutputStream)) { tar =>
+            tar.setBigNumberMode(TarArchiveOutputStream.BIGNUMBER_STAR)
+            tar.setLongFileMode(TarArchiveOutputStream.LONGFILE_GNU)
+            tar.setAddPaxHeadersForNonAsciiNames(true)
+            archive(".tar.gz", tar) { (path, size, mode) =>
+              val entry = new TarArchiveEntry(path)
+              entry.setSize(size)
+              entry.setMode(mode)
+              entry
+            }
+          }
+        }
+      case _ =>
+        BadRequest()
```

Comment:
Return type of `archiveRepository` is `Unit` in signature. So even if `BadRequest` is returned here, it would not be responded to the client.